book/styles.md

```diff
@@ -689,112 +662,232 @@ happen.[^ours-works]
+
+So, in fact, browsers resolve percentages to absolute pixel units
+before storing them in the `style`; it's called a ["computed
+style"][^css-computed]. Of the properties our toy browser supports,
+only `font-size` needs to be computed in this way:
 
-On `TextNode` objects we can do an even simpler trick, since a text
-node never has styles of its own and only inherits from its parent:
+[^css-computed]: Full CSS is a bit more confusing: there are
+[specified, computed, used, and actual values][css-computed], and they
+affect lots of CSS properties besides `font-size`. We're just not
+implementing those other properties in this book.
+
+[css-computed]: https://www.w3.org/TR/CSS2/cascade.html#value-stages
 
 ``` {.python}
-def style(node, rules):
-    if isinstance(node, TextNode):
-        node.style = node.parent.style
+def compute_style(node, property, value):
+    if property == "font-size":
+        if value.endswith("px"):
+            return value
+        elif value.endswith("%"):
+            # ...
+        else:
+            return None
     else:
-        # ...
+        return value
 ```
 
-With all this in place, we can implement text color itself. First,
-let's add a `color` parameter to `DrawText` and pass it to
-`create_text`'s `fill` parameter:
+The case for percentage sizes has to handle a tricky edge case:
+percentage sizes for the root `html` element. In that case the
+percentage is relative to the default font size:[^why-parse]
+
+[^why-parse]: This code has to parse and unparse font sizes because
+    our `style` field stores strings; in a real browser the computed
+    style is stored parsed so this doesn't have to happen.
+
+``` {.python indent=12}
+if node.parent:
+    parent_font_size = node.parent.style["font-size"]
+else:
+    parent_font_size = INHERITED_PROPERTIES["font-size"]
+node_pct = float(value[:-1]) / 100
+parent_px = float(parent_font_size[:-2])
+return str(node_pct * parent_px) + "px"
+```
+
+Now `style` can call `computed_style` any time it reads a property
+value out of a style sheet:
 
 ``` {.python}
-class DrawText:
-    def __init__(self, x1, y1, text, font, color):
-        # ...
-        self.color = color
+def style(node, rules):
+    # ...
+    for selector, body in rules:
+        if not selector.matches(node): continue
+        for property, value in body.items():
+            computed_value = compute_style(node, property, value)
+            if not computed_value: continue
+            node.style[property] = computed_value
+    # ...
+```
 
-    def execute(self, scroll, canvas):
-        canvas.create_text(
-            # ...
-            color=self.color,
-        )
+The loop that handles `style` attributes should likewise call
+`computed_style`. Note that because the `style` function recurses at
+the end of the function, the node's parent already has a `font-size`
+value stored when `compute_style` is called.
+
+::: {.further}
```

Comment:
remove - empty

Reply:
I'm going to fill the two empty ones in with something interesting. Open to suggestions—thinking of talking about rule trees in the "Inheritance" section and promoting the long footnote on pixels v points to the "Font Properties" go further block.

Reply:
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/style-calculation.md doesn't have much..